File: packages/google_mlkit_image_labeling/ios/Classes/GoogleMlKitImageLabelingPlugin.m
#import "GoogleMlKitImageLabelingPlugin.h"
#import <MLKitCommon/MLKitCommon.h>
#import <MLKitImageLabeling/MLKitImageLabeling.h>
#import <MLKitImageLabelingCommon/MLKitImageLabelingCommon.h>
#import <MLKitImageLabelingCustom/MLKitImageLabelingCustom.h>
#import <google_mlkit_commons/GoogleMlKitCommonsPlugin.h>

#if MLKIT_FIREBASE_MODELS
#import <MLKitLinkFirebase/MLKitLinkFirebase.h>
#endif

#define channelName @"google_mlkit_image_labeler"
#define startImageLabelDetector @"vision#startImageLabelDetector"
#define closeImageLabelDetector @"vision#closeImageLabelDetector"
#define manageFirebaseModels @"vision#manageFirebaseModels"

@implementation GoogleMlKitImageLabelingPlugin {
    NSMutableDictionary *instances;
    GenericModelManager *genericModelManager;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:channelName
                                     binaryMessenger:[registrar messenger]];
    GoogleMlKitImageLabelingPlugin* instance = [[GoogleMlKitImageLabelingPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)init {
    self = [super init];
    if (self)
        instances = [NSMutableDictionary dictionary];
    return  self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([call.method isEqualToString:startImageLabelDetector]) {
        [self handleDetection:call result:result];
    } else if ([call.method isEqualToString:closeImageLabelDetector]) {
        NSString *uid = call.arguments[@"id"];
        [instances removeObjectForKey:uid];
        result(NULL);
    } else if ([call.method isEqualToString:manageFirebaseModels]) {
#if MLKIT_FIREBASE_MODELS
        [self manageModel:call result:result];
#else
        result([FlutterError errorWithCode:@"ERROR_MISSING_MLKIT_FIREBASE_MODELS" message:@"You must define MLKIT_FIREBASE_MODELS=1 in your Podfile." details:nil]);
#endif
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)handleDetection:(FlutterMethodCall *)call result:(FlutterResult)result {
    MLKVisionImage *image = [MLKVisionImage visionImageFromData:call.arguments[@"imageData"]];
    
    NSString *uid = call.arguments[@"id"];
    MLKImageLabeler *labeler = [instances objectForKey:uid];
    if (labeler == NULL) {
        NSDictionary *dictionary = call.arguments[@"options"];
        NSString *type = dictionary[@"type"];
        if ([@"base" isEqualToString:type]) {
            MLKImageLabelerOptions *options = [self getDefaultOptions:dictionary];
            labeler = [MLKImageLabeler imageLabelerWithOptions:options];
        } else if ([@"local" isEqualToString:type]) {
            MLKCustomImageLabelerOptions *options = [self getLocalOptions:dictionary];
            labeler = [MLKImageLabeler imageLabelerWithOptions:options];
        } else if ([@"remote" isEqualToString:type]) {
#if MLKIT_FIREBASE_MODELS
            MLKCustomImageLabelerOptions *options = [self getRemoteOptions:dictionary];
            if (options == NULL) {
                FlutterError *error = [FlutterError errorWithCode:@"Error Model has not been downloaded yet"
                                                          message:@"Model has not been downloaded yet"
                                                          details:@"Model has not been downloaded yet"];
                result(error);
                return;
            }
            labeler = [MLKImageLabeler imageLabelerWithOptions:options];
#else
            result([FlutterError errorWithCode:@"ERROR_MISSING_MLKIT_FIREBASE_MODELS" message:@"You must define MLKIT_FIREBASE_MODELS=1 in your Podfile." details:nil]);
#endif
        } else {
            NSString *error = [NSString stringWithFormat:@"Invalid model type: %@", type];
            result([FlutterError errorWithCode:type
                                       message:error
                                       details:error]);
            return;
        }
        instances[uid] = labeler;
    }
    
    [labeler processImage:image
               completion:^(NSArray<MLKImageLabel *> *_Nullable labels,
                            NSError *_Nullable error) {
        if (error) {
            result(getFlutterError(error));
            return;
        } else if (!labels) {
            result(@[]);
        }
        
        NSMutableArray *labelData = [NSMutableArray array];
        for (MLKImageLabel *label in labels) {
            NSDictionary *data = @{
                @"confidence" : @(label.confidence),
                @"index" : @(label.index),
                @"text" : label.text,
            };
            [labelData addObject:data];
        }
        
        result(labelData);
    }];
}

- (MLKImageLabelerOptions *)getDefaultOptions:(NSDictionary *)optionsData {
    NSNumber *conf = optionsData[@"confidenceThreshold"];
    MLKImageLabelerOptions *options = [MLKImageLabelerOptions new];
    options.confidenceThreshold = conf;
    return options;
}

- (MLKCustomImageLabelerOptions *)getLocalOptions:(NSDictionary *)optionsData {
    NSNumber *conf = optionsData[@"confidenceThreshold"];
    NSNumber *maxCount = optionsData[@"maxCount"];
    NSString *path = optionsData[@"path"];
    
    MLKLocalModel *localModel = [[MLKLocalModel alloc] initWithPath:path];
    MLKCustomImageLabelerOptions *options = [[MLKCustomImageLabelerOptions alloc] initWithLocalModel:localModel];
    options.confidenceThreshold = conf;
    options.maxResultCount = maxCount.intValue;
    return options;
}

#if MLKIT_FIREBASE_MODELS
- (MLKCustomImageLabelerOptions *)getRemoteOptions:(NSDictionary *)optionsData {
    NSNumber *conf = optionsData[@"confidenceThreshold"];
    NSNumber *maxCount = optionsData[@"maxCount"];
    NSString *modelName = optionsData[@"modelName"];
    
    MLKFirebaseModelSource *firebaseModelSource = [[MLKFirebaseModelSource alloc] initWithName:modelName];
    MLKCustomRemoteModel *remoteModel = [[MLKCustomRemoteModel alloc] initWithRemoteModelSource:firebaseModelSource];
    
    MLKModelManager *modelManager = [MLKModelManager modelManager];
    BOOL isModelDownloaded = [modelManager isModelDownloaded:remoteModel];
    if (!isModelDownloaded) {
        return NULL;
    }
    
    MLKCustomImageLabelerOptions *options = [[MLKCustomImageLabelerOptions alloc] initWithRemoteModel:remoteModel];
    options.confidenceThreshold = conf;
    options.maxResultCount = maxCount.intValue;
    return options;
}

- (void)manageModel:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *modelTag = call.arguments[@"model"];
    MLKFirebaseModelSource *firebaseModelSource = [[MLKFirebaseModelSource alloc] initWithName:modelTag];
    MLKCustomRemoteModel *model = [[MLKCustomRemoteModel alloc] initWithRemoteModelSource:firebaseModelSource];
    genericModelManager = [[GenericModelManager alloc] init];
    [genericModelManager manageModel:model call:call result:result];
}
#endif

@end

File: packages/google_mlkit_image_labeling/ios/google_mlkit_image_labeling/Classes/GoogleMlKitImageLabelingPlugin.swift
import Flutter
import MLKitCommon
import MLKitVision
import MLKitImageLabeling
import MLKitImageLabelingCommon
import MLKitImageLabelingCustom
import google_mlkit_commons

#if canImport(MLKitLinkFirebase)
import MLKitLinkFirebase
#endif

@objc
public class GoogleMlKitImageLabelingPlugin: NSObject, FlutterPlugin {
  private var instances: [String: ImageLabeler] = [:]
  #if canImport(MLKitLinkFirebase)
  private var genericModelManager: GenericModelManager?
  #endif

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "google_mlkit_image_labeler",
      binaryMessenger: registrar.messenger()
    )
    let instance = GoogleMlKitImageLabelingPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "vision#startImageLabelDetector":
      handleDetection(call: call, result: result)
    case "vision#closeImageLabelDetector":
      if let args = call.arguments as? [String: Any], let uid = args["id"] as? String {
        instances.removeValue(forKey: uid)
      }
      result(nil)
    case "vision#manageFirebaseModels":
      #if canImport(MLKitLinkFirebase)
      manageModel(call: call, result: result)
      #else
      result(FlutterError(
        code: "ERROR_MISSING_MLKIT_FIREBASE_MODELS",
        message: "Add the GoogleMLKit/LinkFirebase pod to your Podfile to use Firebase-hosted models.",
        details: nil
      ))
      #endif
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleDetection(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let imageData = args["imageData"] as? [String: Any],
          let uid = args["id"] as? String else {
      result(FlutterError(code: "invalid_args", message: "Missing arguments", details: nil))
      return
    }
    guard let image = VisionImage.visionImage(from: imageData) else {
      result(FlutterError(code: "invalid_image", message: "Invalid or missing image data", details: nil))
      return
    }

    let labeler: ImageLabeler
    if let existing = instances[uid] {
      labeler = existing
    } else {
      guard let dictionary = args["options"] as? [String: Any] else {
        result(FlutterError(code: "invalid_args", message: "Missing options", details: nil))
        return
      }
      let type = dictionary["type"] as? String ?? "base"
      switch type {
      case "base":
        labeler = ImageLabeler.imageLabeler(options: getDefaultOptions(dictionary))
      case "local":
        guard let options = getLocalOptions(dictionary) else {
          result(FlutterError(code: "invalid_args", message: "Missing path for local model", details: nil))
          return
        }
        labeler = ImageLabeler.imageLabeler(options: options)
      case "remote":
        #if canImport(MLKitLinkFirebase)
        guard let modelName = dictionary["modelName"] as? String, !modelName.isEmpty else {
          result(FlutterError(
            code: "invalid_args",
            message: "Missing or invalid modelName for remote model",
            details: nil
          ))
          return
        }
        if let options = getRemoteOptions(dictionary) {
          labeler = ImageLabeler.imageLabeler(options: options)
        } else {
          result(FlutterError(
            code: "Error Model has not been downloaded yet",
            message: "Model has not been downloaded yet",
            details: "Model has not been downloaded yet"
          ))
          return
        }
        #else
        result(FlutterError(
          code: "ERROR_MISSING_MLKIT_FIREBASE_MODELS",
          message: "Add the GoogleMLKit/LinkFirebase pod to your Podfile to use Firebase-hosted models.",
          details: nil
        ))
        return
        #endif
      default:
        result(FlutterError(
          code: type,
          message: "Invalid model type: \(type)",
          details: "Invalid model type: \(type)"
        ))
        return
      }
      instances[uid] = labeler
    }

    labeler.process(image) { labels, error in
      if let error = error as NSError? {
        result(FlutterError(code: "Error \(error.code)", message: error.domain, details: error.localizedDescription))
        return
      }
      guard let labels = labels else {
        result([])
        return
      }
      let labelData = labels.map { label in
        [
          "confidence": label.confidence,
          "index": label.index,
          "text": label.text
        ] as [String: Any]
      }
      result(labelData)
    }
  }

  private func getDefaultOptions(_ optionsData: [String: Any]) -> ImageLabelerOptions {
    let options = ImageLabelerOptions()
    if let conf = optionsData["confidenceThreshold"] as? NSNumber {
      options.confidenceThreshold = conf
    }
    return options
  }

  private func getLocalOptions(_ optionsData: [String: Any]) -> CustomImageLabelerOptions? {
    guard let path = optionsData["path"] as? String else { return nil }
    let localModel = LocalModel(path: path)
    let options = CustomImageLabelerOptions(localModel: localModel)
    if let conf = optionsData["confidenceThreshold"] as? NSNumber {
      options.confidenceThreshold = conf
    }
    if let maxCount = optionsData["maxCount"] as? NSNumber {
      options.maxResultCount = maxCount.intValue
    }
    return options
  }

  #if canImport(MLKitLinkFirebase)
  private func getRemoteOptions(_ optionsData: [String: Any]) -> CustomImageLabelerOptions? {
    guard let modelName = optionsData["modelName"] as? String else { return nil }
    let firebaseModelSource = FirebaseModelSource(name: modelName)
    let remoteModel = CustomRemoteModel(remoteModelSource: firebaseModelSource)
    guard ModelManager.modelManager().isModelDownloaded(remoteModel) else {
      return nil
    }
    let options = CustomImageLabelerOptions(remoteModel: remoteModel)
    if let conf = optionsData["confidenceThreshold"] as? NSNumber {
      options.confidenceThreshold = conf
    }
    if let maxCount = optionsData["maxCount"] as? NSNumber {
      options.maxResultCount = maxCount.intValue
    }
    return options
  }

  private func manageModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let modelTag = args["model"] as? String else {
      result(FlutterError(code: "invalid_args", message: "Missing model argument", details: nil))
      return
    }
    let firebaseModelSource = FirebaseModelSource(name: modelTag)
    let model = CustomRemoteModel(remoteModelSource: firebaseModelSource)
    let manager = GenericModelManager()
    genericModelManager = manager
    manager.manage(model: model, call: call, result: result)
  }
  #endif
}
